File: objc/main/test/objc/TestScopes.m
#import "TestScopes.h"

@implementation TestScopes

- (void) testSeesEnclosingAttr {
    NSString *templates = @"t(x,y) ::= \"<u()>\"\nu() ::= \"<x><y>\"";
    ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
    [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
    STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
    [group setListener:errors];
    ST *st = [group getInstanceOf:@"t"];
    [st add:@"x" value:@"x"];
    [st add:@"y" value:@"y"];
    NSString *result = [st render];
    NSString *expectedError = @"";
    [self assertEquals:expectedError arg1:[errors description]];
    NSString *expected = @"xy";
    [self assertEquals:expected arg1:result];
}

- (void) testMissingArg {
    NSString *templates = [@"t() ::= \"<u()>\"\n" stringByAppendingString:@"u(z) ::= \"\""];
    ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
    [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
    STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
    [group setListener:errors];
    ST *st = [group getInstanceOf:@"t"];
    NSString *result = [st render];
    NSString *expectedError = [@"context [t] 1:1 passed 0 arg(s) to template u with 1 declared arg(s)" stringByAppendingString:newline];
    [self assertEquals:expectedError arg1:[errors description]];
}

- (void) testUnknownAttr {
    NSString *templates = @"t() ::= \"<x>\"\n";
    ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
    [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
    STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
    [group setListener:errors];
    ST *st = [group getInstanceOf:@"t"];
    NSString *result = [st render];
    NSString *expectedError = [@"context [t] 1:1 attribute x isn't defined" stringByAppendingString:newline];
    [self assertEquals:expectedError arg1:[errors description]];
}

- (void) testArgWithSameNameAsEnclosing {
    NSString *templates = [@"t(x,y) ::= \"<u(x)>\"\n" stringByAppendingString:@"u(y) ::= \"<x><y>\""];
    ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
    [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
    STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
    [group setListener:errors];
    ST *st = [group getInstanceOf:@"t"];
    [st add:@"x" value:@"x"];
    [st add:@"y" value:@"y"];
    NSString *result = [st render];
    NSString *expectedError = @"";
    [self assertEquals:expectedError arg1:[errors description]];
    NSString *expected = @"xx";
    [self assertEquals:expected arg1:result];
    [group setListener:ErrorManager.DEFAULT_ERROR_LISTENER];
}

- (void) testIndexAttrVisibleLocallyOnly {
    NSString *templates = [@"t(names) ::= \"<names:{n | <u(n)>}>\"\n" stringByAppendingString:@"u(x) ::= \"<i>:<x>\""];
    ErrorBuffer *errors = [ErrorBuffer newErrorBuffer];
    [self writeFile:tmpdir arg1:@"t.stg" arg2:templates];
    STGroup *group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"t.stg"]] autorelease];
    [group setListener:errors];
    ST *st = [group getInstanceOf:@"t"];
    [st add:@"names" arg1:@"Ter"];
    NSString *result = [st render];
    [[group getInstanceOf:@"u"].impl dump];
    NSString *expectedError = [@"t.stg 2:11: attribute i isn't defined" stringByAppendingString:newline];
    [self assertEquals:expectedError arg1:[errors description]];
    NSString *expected = @":Ter";
    [self assertEquals:expected arg1:result];
    [group setListener:ErrorManager.DEFAULT_ERROR_LISTENER];
}

@end

File: test/org/stringtemplate/v4/test/TestScopes.java
package org.stringtemplate.v4.test;

import org.junit.Test;
import org.stringtemplate.v4.*;
import org.stringtemplate.v4.misc.*;

import static org.junit.Assert.assertEquals;

public class TestScopes extends BaseTest {
    @Test public void testSeesEnclosingAttr()
    {
        String templates =
            "t(x,y) ::= \"<u()>\"\n" +
            "u() ::= \"<x><y>\"";
        ErrorBuffer errors = new ErrorBuffer();
        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.add("x", "x");
        st.add("y", "y");
        String result = st.render();

        String expectedError = "";
        assertEquals(expectedError, errors.toString());

        String expected = "xy";
        assertEquals(expected, result);
    }

    @Test public void testMissingArg()
    {
        String templates =
            "t() ::= \"<u()>\"\n" +
            "u(z) ::= \"\"";
        ErrorBuffer errors = new ErrorBuffer();
        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        String result = st.render();

        String expectedError = "context [/t] 1:1 passed 0 arg(s) to template /u with 1 declared arg(s)"+newline;
        assertEquals(expectedError, errors.toString());
    }

    @Test public void testUnknownAttr()
    {
        String templates =
            "t() ::= \"<x>\"\n";
        ErrorBuffer errors = new ErrorBuffer();
        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        String result = st.render();

        String expectedError = "context [/t] 1:1 attribute x isn't defined"+newline;
        assertEquals(expectedError, errors.toString());
    }

    @Test public void testArgWithSameNameAsEnclosing()
    {
        String templates =
            "t(x,y) ::= \"<u(x)>\"\n" +
            "u(y) ::= \"<x><y>\"";
        ErrorBuffer errors = new ErrorBuffer();
        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.add("x", "x");
        st.add("y", "y");
        String result = st.render();

        String expectedError = "";
        assertEquals(expectedError, errors.toString());

        String expected = "xx";
        assertEquals(expected, result);
        group.setListener(ErrorManager.DEFAULT_ERROR_LISTENER);
    }

    @Test public void testIndexAttrVisibleLocallyOnly()
    {
        String templates =
            "t(names) ::= \"<names:{n | <u(n)>}>\"\n" +
            "u(x) ::= \"<i>:<x>\"";
        ErrorBuffer errors = new ErrorBuffer();
        writeFile(tmpdir, "t.stg", templates);
        STGroup group = new STGroupFile(tmpdir+"/"+"t.stg");
        group.setListener(errors);
        ST st = group.getInstanceOf("t");
        st.add("names", "Ter");
        String result = st.render();
        group.getInstanceOf("u").impl.dump();

        String expectedError = "t.stg 2:11: implicitly-defined attribute i not visible"+newline;
        assertEquals(expectedError, errors.toString());

        String expected = ":Ter";
        assertEquals(expected, result);
        group.setListener(ErrorManager.DEFAULT_ERROR_LISTENER);
    }

}
